### PR TITLE
feat: add wordwrap for text

### PR DESCRIPTION
**Pull request by markov00**
## Summary
This is an improved version of the current wordWrap function that we have.
It uses the `Intl.Segmenter` or a fallback to split the text into multiple parts and uses the monotonic hill climber to find the right place to trim the word at the right character.
Ellipsis is added always at the end of the last line if more text is present but doesn't with in the provided `maxLineWidth` and `maxLines` parameters.

#### Further improvements

- [ ] handle ellipsis earlier: don't push to a new line the word if we are already at the last line. try to keep and cut the last word even if is longer then the available space
- [ ] improve the layout algorighm to increase the used space: as an example we can trim long words into multi-line only if the available space to render the word is bigger then half of the `maxLineWidth`, this suffers from hyphenation (that we currently don't support [as most browser don't](https://developer.mozilla.org/en-US/docs/Web/CSS/hyphens#browser_compatibility))
- [ ] verify the new line and multi-space omissions and in case allow those
- [ ] Understand if is worth using alternatives algorithms like [SMAWK](https://en.wikipedia.org/wiki/SMAWK_algorithm)

## Issues

See demo story [here](https://ech-e2e-ci--pr-1761-0q7d9yi3.web.app/?path=/story/utils-text--wrap&globals=theme:light)

### Checklist

- [ ] The proper **chart type** label has been added (e.g. `:xy`, `:partition`)
- [ ] The proper **feature** labels have been added (e.g. `:interactions`, `:axis`)
- [ ] The `:theme` label has been added and the `@elastic/eui-design` team has been pinged when there are `Theme` API changes
- [ ] All related issues have been linked (i.e. `closes #123`, `fixes #123`)
- [ ] New public API exports have been added to `packages/charts/src/index.ts`
- [ ] Unit tests have been added or updated to match the most common scenarios
- [ ] The proper documentation and/or storybook story has been added or updated
- [ ] The code has been checked for cross-browser compatibility (Chrome, Firefox, Safari, Edge)
- [ ] Visual changes have been tested with all available themes including `dark`, `light`, `eui-dark` & `eui-light`

### CODE REVIEW

**Comment by markov00** [comment]
buildkite update screenshots

File: storybook/stories/utils/text/1_wrap.story.tsx [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+const fontSize = 24;
+const font: Font = {
+  fontStyle: 'normal',
+  fontFamily: 'sans-serif',
+  fontVariant: 'normal',
+  fontWeight: 500,
+  textColor: 'red',
+};
+const fontStyle = cssFontShorthand(font, fontSize);
+const defaultText =
```
Comment (nickofthyme) [inline storybook/stories/utils/text/1_wrap.story.tsx:25]: Not sure where the edge case is caused in code but between a width of `259px` and `265px` it incorrectly creates a blank line where the `mongoloadgendecoblue58d844d55c-9c24dtip` string is too long to fill a single line. I'm not sure how we can test all these types of edge cases sufficiently without rendering all or some widths conditionally.


Screen Recording 2022-07-25 at 08 28 42 PM

File: storybook/stories/utils/text/1_wrap.story.tsx [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+  'Bacon ipsum dolor amet mongoloadgendecoblue58d844d55c-9c24dtip flank kielbasa. Pork strip steak jowl chuck filet mignon, burgdoggen kevin tail.';
+
+export const Example = () => {
+  const [maxLineWidth, setMaxLineWidth] = useState(300);
+  const [maxLines, setMaxLines] = useState(3);
```
Comment (nickofthyme) [inline storybook/stories/utils/text/1_wrap.story.tsx:30]: Why no use knobs here? I'm fine if that's what you want but this will prevent us from programmatically rendering stories in VRT.

File: packages/charts/src/utils/text/wrap.ts [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+  locale: string[],
+): (text: string) => { segment: string; index: number; isWordLike?: boolean }[] {
+  if ('Segmenter' in Intl) {
+    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+    //@ts-ignore
+    const fn = new Intl.Segmenter(locale, { granularity });
+    return (text: string) => Array.from(fn.segment(text));
+  } else {
+    return function (text: string) {
+      return text.split(' ').map((segment, index) => {
```
Comment (nickofthyme) [inline packages/charts/src/utils/text/wrap.ts:25]: Having more than two spaces here could cause issues. Would not be as simple as adding regex like `/\w+/` because we'd need to know the space count to put back. I don't think it's a big deal but certainly and edge case. The current logic just destroys any multiple of spaces.

Utils Text - Wrap ⋅ Storybook 2022-07-25 at 8 33 30 PM

File: packages/charts/src/utils/text/wrap.ts [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+        breakupWords.length > 0 ? measure(breakupWords[breakupWords.length - 1], font, fontSize, lineHeight).width : 0;
+    } else {
+      if (lines.length === 0) {
+        lines.push(segment.segment.trimStart());
+        currentLineWidth += measure(segment.segment.trimStart(), font, fontSize, lineHeight).width;
+      } else {
+        lines[lines.length - 1] += segment.segment;
+        currentLineWidth += segment.width;
+      }
+    }
```
Comment (nickofthyme) [inline packages/charts/src/utils/text/wrap.ts:78]: In the blank line case mentioned above, we are passing an array of an empty string array (i.e. `['']`) as `breakupWords` but we do not do get the `['']` when the width accommodates the word length. My guess is some logic in there is the cause.

```
wrap.ts:70 [''] # <-- this seems 🐠y
wrap.ts:70 (2) ['mongoloadgendecoblue', '58d844d55c']
wrap.ts:70 ['flank']
wrap.ts:70 ['strip']
wrap.ts:70 ['filet']
wrap.ts:70 ['burgdoggen']
```

**Review by nickofthyme — COMMENTED** [review]
Most of the logic looks sounds to me for what I understand. But it would be nice to do a code walkthrough once @monfera returns to identify possible edge cases.

In the meantime I think it's fine to merge this. :shipit:

File: storybook/stories/utils/text/1_wrap.story.tsx (hunk 2 above)
Comment (markov00) [reply storybook/stories/utils/text/1_wrap.story.tsx:30]: I will revert that to knobs whenever we will need the programmatic rendering. Using the in-story one is way more efficient and faster to play/test then with the denounced knobs

File: packages/charts/src/utils/text/wrap.ts (hunk 3 above)
Comment (markov00) [reply packages/charts/src/utils/text/wrap.ts:25]: exactly, the current logic is to remove subsequent white spaces and also removes new lines. In our context (a very limited amount of space where we want to fit as much text as possible), rendering multiple spaces or new lines is something that we like to avoid and that's why I'm removing those elements from the initial text.

**Review by nickofthyme — APPROVED** [review]

**Comment by markov00** [comment]
buildkite update screenshots

**Review by monfera — COMMENTED** [review]
The functionality is very nice! The implementation could be made less convoluted, though some of the suggestions may reflect code/intent misreading, or unreasonable time for rework. None of my comments is a blocker

Also, let's please add a gh issue for the eventual unification with the multirow stuff in partition charts, as it would be useful there too, and also to be DRY

File: packages/charts/src/utils/text/wrap.ts [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+  } else {
+    return function (text: string) {
+      return text
+        .split(' ')
+        .reduce<{ segment: string; index: number; isWordLike?: boolean }[]>((acc, segment, index, array) => {
+          const currentSegment = { segment, index: index === 0 ? 0 : acc[acc.length - 1].index + 1, isWordLike: true };
+          acc.push(currentSegment);
+          // adding space to simulate the same behaviour of the segmenter in firefox
+          if (index < array.length - 1) {
+            acc.push({ segment: ' ', index: currentSegment.index + segment.length, isWordLike: false });
```
[the remark this answers is not in the document]
Comment (markov00) [reply packages/charts/src/utils/text/wrap.ts:32]: The only reason here is to have a similar result to the one returned by the `Intl.Segmenter`.
In particular, the Segmenter returns separate elements for each whitespace, punctuation marks etc. Our naive, fallback, method returns only a set of strings that are separated by the space character `' '`, we need to add the space after those to have the correct set of characters of the initial text

File: packages/charts/src/utils/text/wrap.ts [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+  fontSize: number,
+  maxLineWidth: number,
+  maxLines: number,
+): string[] {
+  if (maxLines <= 0) {
+    return [];
+  }
+  const segmenter = textSegmenter('word', []);
+  // remove new lines and multi-spaces.
+  const cleanedText = text.replace(/\n/g, ' ').replace(/ +(?= )/g, '');
```
[the remark this answers is not in the document]
Comment (markov00) [reply packages/charts/src/utils/text/wrap.ts:56]: yea, actually I haven't thought about removing those exotic white spaces. It can be improved in the long run

File: packages/charts/src/utils/text/wrap.ts [hunk 7]
```diff
@@ -0,0 +1,163 @@
...
+    ...d,
+    width: measure(d.segment, font, fontSize, lineHeight).width,
+  }));
+
+  const ellipsisWidth = measure(ELLIPSIS, font, fontSize).width;
+  const lines: string[] = [];
+  let currentLineWidth = 0;
+  for (const segment of segments) {
+    // the word is longer then the available space and is not a space
+    if (currentLineWidth + segment.width > maxLineWidth && segment.segment.trimStart().length > 0) {
```
[the remark this answers is not in the document]
Comment (markov00) [reply packages/charts/src/utils/text/wrap.ts:68]: You are probably right, I'm pretty sure I've add that here for a specific reason, but I will double check that.

The `segment` name is because we follow the current `Intl.Segmenter` API that returns an array of `{segment, index, isWordLike}`

File: packages/charts/src/utils/text/wrap.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (monfera) [reply packages/charts/src/utils/text/wrap.ts:32]: Got it, it's good. A potential alternative is to directly eliminate the spacers that `Intl.Segmenter` returns, so then both tracks can just deal with the info and not carry around the noise (zero info content). Ofc it's fine either way

File: packages/charts/src/utils/text/wrap.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (monfera) [reply packages/charts/src/utils/text/wrap.ts:68]: Thanks, poor `Intl` API uses it for both the object and the property indeed 🙈

File: packages/charts/src/utils/text/wrap.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (markov00) [reply packages/charts/src/utils/text/wrap.ts:32]: Without dealing with everything else the `Segmenter` returns like punctuation marks, I don't know if it is possible to reconstruct the original phrase.

File: packages/charts/src/utils/text/wrap.ts [hunk 8]
```diff
@@ -0,0 +1,163 @@
...
+): (text: string) => { segment: string; index: number; isWordLike?: boolean }[] {
+  if ('Segmenter' in Intl) {
+    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+    //@ts-ignore
+    const fn = new Intl.Segmenter(locale, { granularity });
+    return (text: string) => Array.from(fn.segment(text));
+  } else {
+    return function (text: string) {
+      return text
+        .split(' ')
```
[the remark this answers is not in the document]
Comment (markov00) [reply packages/charts/src/utils/text/wrap.ts:26]: Exactly, right now, we are cleaning (removing multi-space and new lines for now) the text before splitting it by spaces. This can be improved in the future and we can clean up better or we can keep that and consider various split methodologies.

File: packages/charts/src/utils/text/wrap.ts [hunk 9]
```diff
@@ -0,0 +1,163 @@
...
+  measure: TextMeasure,
+  lineHeight: number,
+  font: Font,
+  fontSize: number,
+  maxLineWidth: number,
+  maxLines: number,
+): string[] {
+  if (maxLines <= 0) {
+    return [];
+  }
```
[the remark this answers is not in the document]
Comment (markov00) [reply packages/charts/src/utils/text/wrap.ts:53]: actually no, is not taking care about the maxLines until the end of the function. It basically go through the text measurement and segmentization and then just at the end checked the `maxLines`.
I prefer to keep this at the beginning to cut all the unnecessary work at the start

File: packages/charts/src/utils/text/wrap.ts [hunk 10]
```diff
@@ -0,0 +1,163 @@
...
+      currentLineWidth =
+        breakupWords.length > 0 ? measure(breakupWords[breakupWords.length - 1], font, fontSize, lineHeight).width : 0;
+    } else {
+      if (lines.length === 0) {
+        lines.push(segment.segment.trimStart());
+        currentLineWidth += measure(segment.segment.trimStart(), font, fontSize, lineHeight).width;
+      } else {
+        lines[lines.length - 1] += segment.segment;
+        currentLineWidth += segment.width;
+      }
```
[the remark this answers is not in the document]
Comment (markov00) [reply packages/charts/src/utils/text/wrap.ts]: Totally, I've cleaned up this and removed the ifs 799001c55146775d1850d288246713955fab6e96

File: packages/charts/src/utils/text/wrap.ts [hunk 11]
```diff
@@ -0,0 +1,163 @@
...
+        measure,
+        lineHeight,
+        font,
+        fontSize,
+        maxLineWidth,
+        Infinity,
+      );
+      if (breakupWords.length === 0) {
+        break;
+      }
```
[the remark this answers is not in the document]
Comment (markov00) [reply packages/charts/src/utils/text/wrap.ts]: I've added a comment here 799001c55146775d1850d288246713955fab6e96
It can be definitely improved but I think there is the need to add other type of checks along the lines to avoid that one

File: packages/charts/src/utils/text/wrap.ts [hunk 12]
```diff
@@ -0,0 +1,163 @@
...
+  maxLines: number, // just for optimization
+) {
+  const lines: string[] = [];
+  let currentWordSection = word;
+  while (maxLines > lines.length && currentWordSection.length > 0) {
+    const line = breakLongWord(currentWordSection, measure, lineHeight, font, fontSize, maxLineWidth);
+    if (line.length === 0) {
+      break;
+    } else {
+      lines.push(line[0].trimStart());
```
[the remark this answers is not in the document]
Comment (markov00) [reply packages/charts/src/utils/text/wrap.ts]: Words can contains trimmables, or better the segmenter returns trimmables.
What we want is to remove white spaces at the beginning of each line.

File: packages/charts/src/utils/text/wrap.ts [hunk 13]
```diff
@@ -0,0 +1,163 @@
...
+  }
+
+  if (maxCharsInLine < word.length) {
+    return [word.slice(0, maxCharsInLine)];
+  }
+
+  if (Number.isNaN(maxCharsInLine)) {
+    return [];
+  }
+  return [word];
```
[the remark this answers is not in the document]
Comment (monfera) [reply packages/charts/src/utils/text/wrap.ts]: Closing this one too, to circle back to in the future
